### PR TITLE
feat(image-plugin): Make it possible to find elements inside of other elements

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Proposed changes

https://github.com/appium/java-client/issues/1891

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

### CODE REVIEW

File: packages/images-plugin/test/unit/image-element.spec.js [hunk 1]
```diff
@@ -48,7 +48,7 @@ describe('ImageElement', function () {
   describe('.asElement', function () {
     it('should get the webdriver object representation of the element', function () {
       const el = new ImageElement(defTemplate, defRect);
-      el.asElement('ELEMENT').ELEMENT.should.match(/^appium-image-el/);
+      el.asElement().ELEMENT.should.match(/^appium-image-el/);
```
Comment (jlipps) [inline packages/images-plugin/test/unit/image-element.spec.js:51]: should we also verify the w3c element identifier key value is correct? i know this wasn't checked before but seems particularly important now that we only support w3c.

**Review by jlipps — APPROVED** [review]

File: packages/images-plugin/test/unit/image-element.spec.js (hunk 1 above)
Comment (mykola-mokhnach) [reply packages/images-plugin/test/unit/image-element.spec.js]: changed to use unwrapElement helper

**Comment by mykola-mokhnach** [comment]
@jlipps Could you please check the PR one more time?

I did some more changes. I've also replaced jimp with sharp for opencv module (the first one was failing to decode some pictures with weird errors).

I would like to get rid of jimp completely for this plugin in favour of sharp and it would probably be a good idea to do the same in appium/support. I believe @boneskull will also be happy about that

File: packages/images-plugin/lib/finder.js [hunk 2]
```diff
@@ -1,113 +1,60 @@
...
+   * @param {number} maxSize
    */
-  constructor(driver, maxSize = MAX_CACHE_SIZE_BYTES) {
-    this.driver = driver;
+  constructor(maxSize = MAX_CACHE_SIZE_BYTES) {
     this.imgElCache = new LRU({
       max: MAX_CACHE_ITEMS,
       maxSize,
-      sizeCalculation: (el) => el.template.length,
+      sizeCalculation: (el) => el.template.length + (el.matchedImage?.length ?? 0),
```
Comment (jlipps) [inline packages/images-plugin/lib/finder.js:47]: ah so here's the source of the memory leak? nice find!

**Review by jlipps — APPROVED** [review]

**Comment by jlipps** [comment]
we should probably release this as a major version bump

File: packages/images-plugin/lib/finder.js (hunk 2 above)
Comment (mykola-mokhnach) [reply packages/images-plugin/lib/finder.js:47]: no, it is not, but this is anyway more precise than before

**Comment by mykola-mokhnach** [comment]
> we should probably release this as a major version bump

why so? I did not do any breaking changes

**Comment by jlipps** [comment]
drivers used with the plugin must now implement `getWindowRect` instead of `getWindowSize`

**Comment by mykola-mokhnach** [comment]
> drivers used with the plugin must now implement `getWindowRect` instead of `getWindowSize`

Ok, let me put the getWindowSize as a fallback

**Review by jlipps — APPROVED** [review]
OK great!

**Pull request merged**